npm/pkg/controlplane/translation/parseSelector.go

```diff
@@ -238,3 +238,164 @@ func parseSelector(selector *metav1.LabelSelector) (labels []string, vals map[st
+	if !include {
+		setNameWithOp = "!" + setName
+	}
+
+	// in case setNameWithOp exists in a set, do not need to add it.
+	if _, exist := ps.labelSet[setNameWithOp]; exist {
+		return
+	}
+
+	ls := labelSelector{
+		include: include,
+		settype: setType,
+		setName: setName,
+		members: members,
+	}
+
+	ps.labelSelectors = append(ps.labelSelectors, ls)
+	ps.labelSet[setNameWithOp] = struct{}{}
+}
+
+// parseNSSelector parses namespaceSelector and returns slice of labelSelector object
+// which includes operator, setType, ipset name and always nil members slice.
+// Member slices is always nil since parseNSSelector function is called
+// after FlattenNameSpaceSelector function is called, which guarantees
+// there is no matchExpression with multiple values.
+// TODO: good to remove this dependency later if possible.
+func parseNSSelector(selector *metav1.LabelSelector) []labelSelector {
+	// TODO(jungukcho): This will not happen
+	if selector == nil {
+		return []labelSelector{}
+	}
+
+	parsedSelectors := newParsedSelectors()
+
+	// #1. All namespaces case
+	if len(selector.MatchLabels) == 0 && len(selector.MatchExpressions) == 0 {
+		parsedSelectors.addSelector(true, ipsets.KeyLabelOfNamespace, util.KubeAllNamespacesFlag)
+		return parsedSelectors.labelSelectors
+	}
+
+	// #2. MatchLabels
+	for matchKey, matchVal := range selector.MatchLabels {
+		// matchKey + ":" + matchVal (can be empty string) case
+		setName := util.GetIpSetFromLabelKV(matchKey, matchVal)
+		parsedSelectors.addSelector(true, ipsets.KeyValueLabelOfNamespace, setName)
+	}
+
+	// #3. MatchExpressions
+	for _, req := range selector.MatchExpressions {
+		var setName string
+		var setType ipsets.SetType
+		switch op := req.Operator; op {
+		case metav1.LabelSelectorOpIn, metav1.LabelSelectorOpNotIn:
+			// "(!) + matchKey + : + matchVal" case
+			setName = util.GetIpSetFromLabelKV(req.Key, req.Values[0])
+			setType = ipsets.KeyValueLabelOfNamespace
+		case metav1.LabelSelectorOpExists, metav1.LabelSelectorOpDoesNotExist:
+			// "(!) + matchKey" case
+			setName = req.Key
+			setType = ipsets.KeyLabelOfNamespace
+		}
+
+		noNegativeOp := (req.Operator == metav1.LabelSelectorOpIn) || (req.Operator == metav1.LabelSelectorOpExists)
+		parsedSelectors.addSelector(noNegativeOp, setType, setName)
+	}
+
+	return parsedSelectors.labelSelectors
+}
+
+// parsePodSelector parses podSelector and returns slice of labelSelector object
+// which includes operator, setType, ipset name and its members slice.
+// Members slice exists only if setType is only NestedLabelOfPod.
+func parsePodSelector(selector *metav1.LabelSelector, nsInPod string) []labelSelector {
+	// TODO(jungukcho): This will not happen
+	if selector == nil {
+		return []labelSelector{}
+	}
+
+	parsedSelectors := newParsedSelectors()
+
+	// #1. All pods in the nsInPod namespace
+	if len(selector.MatchLabels) == 0 && len(selector.MatchExpressions) == 0 {
+		parsedSelectors.addSelector(true, ipsets.Namespace, nsInPod)
```

Comment:
would we want to add the namespace selector for all pods? Why just conditionally? Again, sorry that I didn't wonder about this in your first PR.

If we want to for all, then we can remove the return statement for this situation since the loops will do nothing when lengths are 0.

Reply:
Sorry. I am not clear about question.
Code is for the case when  
```shell
 podSelector: {}
```

Reply:
Seem to have some issues. Will walk through the code again and update you.

Reply:
Good catch! I fixed it.